### PR TITLE
Fix wonky window dragging when using fancy tgui in some special cases

### PR DESCRIPTION
**Pull request by stylemistake**
## About The Pull Request

This pull request fixes the Issue #44038, where TGUI windows get stuck at the bottom of the screen in this very special case:
- Windows 10 taskbar is not located at the bottom of the screen (in my case, taskbar was on top);
- Fancy UI is enabled.

This is because BYOND uses and reports only the "usable" screen resolution instead of an actual physical resolution _[citation needed]_, and therefore window coordinate space is offset by 40px or more (a size of a Windows 10 taskbar), which causes that rapid uncontrolled propulsion of TGUI windows away from the taskbar.

I also took the liberty to refactor some of the code in 'util/byond' to my liking and adding a few helper methods, including `runCommand(command)` and `callByond(url, params) -> Promise`.

Promise-based interface is a first step to start using async/await and making more dynamic calls in tgui.

## Why It's Good For The Game

Window dragging is now more robust with the (currently default) fancy TGUI.

Code accounts for the screen space offset between BYOND and the in-game browser.

## Changelog
:cl:
fix: fixed wonky window dragging when using fancy tgui in some special cases
/:cl:

### CODE REVIEW

**Comment by optimumtact** [comment]
Nice catch

**Comment by JJRcop** [comment]
how well does ractive and/or gulp handle out of band promises? is it a non-issue?

**Comment by stylemistake** [comment]
@JJRcop what do you mean by out of band? Ractive handles updates through "this.set" very well, because those values are internally observed for changes. Only issue is a small memory leak, where i make those Promises, but I highly doubt this will become an issue.

File: tgui/src/util/byond.js [hunk 1]
```diff
@@ -1,16 +1,70 @@
 const encode = encodeURIComponent
 
-// Helper to generate a BYOND href given 'params' as an object (with an optional 'url' for eg winset).
-export function href (params = {}, url = '') {
-  return `byond://${url}?` + Object.keys(params).map(key => `${encode(key)}=${encode(params[key])}`).join('&')
-}
+// Helper to generate a BYOND href given 'params' as an object
+// (with an optional 'url' for eg winset).
+export const href = (url, params = {}) => {
```
Comment (SpaceManiac) [inline tgui/src/util/byond.js]: is there something wrong with `function`?

Comment (stylemistake) [reply tgui/src/util/byond.js]: It's a legacy syntax with multiple shortcomings (such as hoisting) and it carries redundant features (such as making classes, for which we have an actual "class" keyword), so I just prefer to not use it at all.

With arrow functions, you can:
- Do currying out of the box, like a => b => c
- Omit returns to save space
- Looks syntactically cleaner and unifies the way you declare variables.

Comment (SpaceManiac) [reply tgui/src/util/byond.js]: I don't have anything against arrow functions but we babel this all the way back to IE8 anyways so I'm not sure if it ends up being better in the end

Comment (stylemistake) [reply tgui/src/util/byond.js]: Babel is smart enough to figure out simplest ES5 forms, so no issue there.

I checked with the compiler, e.g.:
```js
// original
const x = value => value + 1
// transformed
var x = function x(value) { ... }
```

**Comment by optimumtact** [comment]
tgui needs recompiling to nuke the merge conflict

**Comment by stylemistake** [comment]
@optimumtact rebased on top of current master, with recompiled bundles.

**Pull request merged**